proposals/records/005-sharding.md

```diff
@@ -0,0 +1,197 @@
+of each index type involves many managed files, threads, etc and can be
+quite complex.
+
+Readers and Writers of a shard must be on the same physical machine for a shard.
+Right now, that also means a single Kubernetes Pod(part of stateful set).
+
+While we can "rebalance" nodes to some extent, we don't have good metrics
+to really know how we should rebalance. Disk size is not a good metric to
+know usage of shard.
+
+Moreover, moving a shard is currently tedious. The only implemented option for us
+is to reindex a shard on a new shard and cutover the reference of the node the shard is on.
+Another option could be to copy all the files; however, this is also slow, error prone,
+requires locks and places coupling on the ingest component and our IndexNodes.
+
+Overview of current problems:
+- IndexNode is managing multiple types of indexes seperately. High overhead.
+- CPU tied to disk and difficult to move shards around
+- Can not add replicas for an existing KB's shards
+- Coupling with shard management and ingest
+
+
+## Proposed Solution
+
+The proposed solution is oriented around providing a unified index to simplify
+how we store shard data on the IndexNode.
+
+With a unified index implementation, it is feasible to provide more
+capabilities around sharding(physical read replicas, rebalancing).
+
+
+### Unified Index
+
+By "unified index," we are referring to an index that stores all data
+in a single disk format that hnsw and bm25 index types utilize.
+
+An important aspect of this unified index design is that it also provides
+segmented storage. By segmented storage, we mean that each write should
+produce a new segment. That new segment will work along with other existing
+segments to provide the shard's full dataset.
+
+[more details on the implementation provided elsewhere]
+
+#### Breaking changes
+
+- Split from tantivy: Tantivy is not compatible with this approach.
+- No full entity graph implementation built in
+- No advanced query support
+
+
+### Replication
+
+Each node will have a commit log to track changes through time,
+communicate changes and track replication to read replicas.
+
+- Commit log is append only file which tracks all shard segments
+  and their files. This is then used for replication to read replicas.
+- Writers are responsible for replicating data to readers consistently
+    - 2 phase "commit". First write files everywhere. Second, "turn on" new files.
+    - Having writers responsible for replication allows us to avoid needing
+      to implement something like raft to coordinate read replica state
+- Each node on the system will have a configured set of physical read replicas
+- [OPTIONAL] Readers are responsible for merging segments
+
+
+### Index coordinator
+
+The Index Coordinator is the component responsible for managing the
+IndexNode cluster.
+
+The Index coordinator is the only component that should "know" about
```

Comment:
What happens if the Index Coordinator goes down? Will we have multiple replicas?

Reply:
we'll have multiple replicas

Reply:
Do you have any proposal for replication and partitioning? Something similar as the logic-physical shard partitioning we have right now?

Reply:
I'm not sure what you mean.

Read replicas are tied to writers that are assigned to. That's it. This proposal does not try to decouple read shards from write shards